### PR TITLE
Fix queryByRaft readIndex failed

### PR DESCRIPTION
**Pull request by Linary**
ReadIndex request rejected because leader has not committed any log entry at its term

Change-Id: I6f5c6fd3770da732caf0c6acd41f0441aa4acb55

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 1]
```diff
@@ -229,12 +237,25 @@ public void run(Status status, long index, byte[] reqCtx) {
...
-        try {
-            return closure.waitFinished();
-        } catch (Throwable t) {
-            throw new BackendException("Failed to query", t);
+        };
+        for (int i = 0; i < RaftSharedContext.QUERY_RETRY_TIMES; i++) {
+            this.node().node().readIndex(BytesUtil.EMPTY_BYTES,
+                                         readIndexClosure);
+            try {
+                return closure.waitFinished();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:245]: rename StoreClosure closure to future

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 2]
```diff
@@ -215,12 +216,19 @@ private Object queryByRaft(Query query, Function<Object, Object> func) {
...
             public void run(Status status, long index, byte[] reqCtx) {
                 if (status.isOk()) {
                     closure.complete(status, () -> func.apply(query));
+                } else if (status.getRaftError() == RaftError.EAGAIN) {
+                    LOG.warn("Failed to execute query {} because of leader " +
+                             "has not committed any log entry at its term",
+                             query);
+                    closure.failure(status, new RaftException(
+                            "Failed to execute query '%s' because of leader " +
+                            "has not committed any log entry at its term",
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:230]: move to "status.getRaftError() == RaftError.EAGAIN" block to line 247, just throw exception in else block

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 3]
```diff
@@ -229,12 +237,25 @@ public void run(Status status, long index, byte[] reqCtx) {
...
+                } catch (InterruptedException ex) {
+                    throw new BackendException("Try to sleep a while for " +
+                                               "querying is interrupted");
+                }
+            } catch (Throwable t) {
+                throw new BackendException("Failed to execute query", t);
+            }
         }
+        throw new BackendException("Failed to query in the case of %s " +
+                                   "retries times");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:258]: collect last error message like "retries times with error: %s"

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java (hunk 1 above)
Comment (Linary) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java]: other place used closure

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 4]
```diff
@@ -229,11 +235,12 @@ public void run(Status status, long index, byte[] reqCtx) {
...
                 }
             }
-        });
+        };
+        this.node().node().readIndex(BytesUtil.EMPTY_BYTES, readIndexClosure);
         try {
             return closure.waitFinished();
         } catch (Throwable t) {
-            throw new BackendException("Failed to query", t);
+            throw new BackendException("Failed to execute query", t);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:243]: move LOG.warn at line 231 here

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 5]
```diff
@@ -246,6 +274,12 @@ public void onCreated(PeerId peer) {
             LOG.info("The node {} replicator has created", peer);
         }
 
+        @Override
+        public void onHeartbeated(PeerId peer) {
+            LOG.info("The node {} replicator has heartbeated", peer);
+            started = true;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:280]: RaftNode.this.started

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 6]
```diff
@@ -266,6 +300,7 @@ private boolean isWriteBufferOverflow(Status status) {
         @Override
         public void onDestroyed(PeerId peer) {
             LOG.warn("The node {} prepare to offline", peer);
+            started = false;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:303]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 7]
```diff
@@ -165,6 +175,24 @@ protected void waitLeaderElected() {
                   this.group(), WAIT_LEADER_TIMEOUT);
     }
 
+    protected boolean waitHeartbeated() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:178]: rename waitStarted

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 8]
```diff
@@ -165,6 +175,24 @@ protected void waitLeaderElected() {
...
+        // Raft election:heartbeat timeout factor
+        int factor = this.node.getRaftOptions().getElectionHeartbeatFactor();
+        int heartbeatInterval = electionTimeout / factor;
+        for (int i = 0; i < RaftSharedContext.QUERY_RETRY_TIMES; i++) {
+            if (this.started) {
+                break;
+            }
+            try {
+                Thread.sleep(heartbeatInterval);
+            } catch (InterruptedException ex) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:189]: rename ex to e

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 9]
```diff
@@ -165,6 +175,24 @@ protected void waitLeaderElected() {
...
+                throw new BackendException("Waiting heartbeated is interrupted");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:190]: add cause param

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 10]
```diff
@@ -212,11 +220,9 @@ private Object queryByRaft(Query query, Function<Object, Object> func) {
             return func.apply(query);
         }
 
-        this.node().waitLeaderElected();
         StoreCommand command = new StoreCommand(StoreAction.QUERY);
         StoreClosure closure = new StoreClosure(command);
-        RaftNode raftNode = this.node();
-        raftNode.node().readIndex(BytesUtil.EMPTY_BYTES, new ReadIndexClosure() {
+        ReadIndexClosure readIndexClosure = new ReadIndexClosure() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:225]: rename StoreClosure closure to future

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 11]
```diff
@@ -86,6 +86,14 @@ public boolean isSchemaStore() {
     public synchronized void open(HugeConfig config) {
         this.store.open(config);
         this.initRaftNodeIfNeeded();
+
+        RaftNode node = this.node();
+        node.waitLeaderElected();
+        if (node.node().isLeader()) {
+            if (!node.waitHeartbeated()) {
+                throw new BackendException("Wait raft node heartbeat failed");
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:94]: Failed to wait raft node heartbeat

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java [hunk 12]
```diff
@@ -86,6 +86,14 @@ public boolean isSchemaStore() {
     public synchronized void open(HugeConfig config) {
         this.store.open(config);
         this.initRaftNodeIfNeeded();
+
+        RaftNode node = this.node();
+        node.waitLeaderElected();
+        if (node.node().isLeader()) {
+            if (!node.waitHeartbeated()) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftBackendStore.java:93]: add a common waitFor(future, timeout) method

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java [hunk 13]
```diff
@@ -199,30 +194,44 @@ public boolean onSnapshotLoad(SnapshotReader reader) {
...
+    @Override
+    public void onStartFollowing(LeaderChangeContext ctx) {
+        LOG.info("The node {} become to follower", this.node.nodeId());
+        this.node.onElected(true);
+        super.onStartFollowing(ctx);
+    }
+
+    @Override
+    public void onStopFollowing(LeaderChangeContext ctx) {
+        LOG.info("The node {} become to follower", this.node.nodeId());
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/StoreStateMachine.java:224]: abdicated from follower

File: hugegraph-core/src/main/java/com/baidu/hugegraph/task/ServerInfoManager.java [hunk 14]
```diff
@@ -337,7 +335,7 @@ private HugeServerInfo serverInfo(Id server) {
     }
 
     private HugeServerInfo removeSelfServerInfo() {
-        if (this.initialized()) {
+        if (this.graph.initialized()) {
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/task/ServerInfoManager.java:338]: should use this.call() if graph.initlalized() will open tx (ensure graph tx closed when graph close)

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 15]
```diff
@@ -138,31 +153,73 @@ public Object submitAndWait(StoreCommand command, StoreClosure closure) {
...
-
-        int consumeTime = WAIT_LEADER_TIMEOUT;
-        int sleepInterval = 100;
-        while (consumeTime > 0) {
-            PeerId leaderId = this.node.getLeaderId();
-            if (leaderId != null) {
-                return;
-            } else {
+        long beginTime = System.currentTimeMillis();
+        int internalTimeout = 3000;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:168]: define const interval

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java [hunk 16]
```diff
@@ -138,31 +153,73 @@ public Object submitAndWait(StoreCommand command, StoreClosure closure) {
...
+            public void run(Status status, long index, byte[] reqCtx) {
+                if (status.isOk()) {
+                    RaftNode.this.started = true;
+                } else {
+                    RaftNode.this.started = false;
+                }
+            }
+        };
+        long beginTime = System.currentTimeMillis();
+        int internalTimeout = 3000;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:202]: define

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java (hunk 8 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/raft/RaftNode.java:210]: address

**Pull request merged**